about format: User display names and avatars

sgwilym opened the pull request:
We've had these informally specified for a long time, and it's time to start making them formal.

This PR only specifies how display names and avatars should be formatted on an Earthstar share. Are there other critical fields which belong in the first version of this format? Perhaps it's worth specifying a way to have freeform fields on a profile, as is available on Mastodon profiles?

Ideas and collaboration welcome in this PR!

sgwilym committed:
Add basic about spec (display name + avatar)

parnikkapore commented on formats/about/SPEC_1.0.md:
```diff
@@ -0,0 +1,61 @@
+
+Therefore, each documents following this format is prefixed with the following:
+
+`/about/1.0/~{KEYPAIR_ADDRESS}`
```
So the tilde directory doesn't need to be at top level after all - nice!

parnikkapore commented on formats/about/SPEC_1.0.md:
````diff
@@ -0,0 +1,61 @@
+`/about/1.0/~{KEYPAIR_ADDRESS}`/avatar.{IMAGE_EXTENSION}
+```
+
+where `IMAGE_EXTENSION` is either `gif`, `jpg`, or `png`, corresponding to the format of the data used with this document's attachment.
````
Only one such document may be present, perhaps? Also, is a client supposed to try all three in order?

sgwilym replied:
Oh, right. This doesn't stop someone from setting all three.

For simplicity's sake, I'd love for clients to be able to generate a path for the avatar pretty deterministically. 

On the other hand, I didn't want to wade into "let's choose one image format for all time". Then either users have to convert image formats, or this conversion needs to be done by clients.

Maybe falling through the different extensions is the best. What do you think?

achou11 replied:
I'd lean towards not limiting the image format here if possible for a couple of reasons:

1. Image formats can change and improve. Don't have any specific examples but there's a chance that newly introduced formats can be preferable, depending on the implementers context.

2. It feels like more a concern for the client to decide and worry about the image format that's being used. This way there's more flexibility for them to design their application as they see fit. As long as there's a path that matches `avatar.*`, that should be enough.

parnikkapore commented on formats/about/SPEC_1.0.md:
```diff
@@ -0,0 +1,61 @@
+
+- The `text` property of the document MUST be a textual description of the avatar image's contents, for use with screenreaders etc.
+- The attachment of the document MUST be an image file.
+	- The file size of the image MUST be less than or equal to 500,000 bytes.
+	- The dimensions of the image MUST NOT exceed 512 pixels on either side.
```
I'm not sure if specifying constraints on the image file's contents is in scope for this spec, especially as a MUST

sgwilym replied:
I'm not an experienced spec writer by any means, so honest question: why would the image file's contents fall outside the scope of this spec?

My intention was that clients would be able to fairly easily be able to predict how to present avatars, and not have to handle weird cases like a super tall image.

achou11 replied:
Agree that I think specifying these constraints seems to be out of scope, for reasons similar to what I provided in https://github.com/earthstar-project/application-formats/pull/1#discussion_r1054428993

parnikkapore commented:
To elaborate: I think that this depends on whether the spec is meant to be there unmodified for a long time (like an RFC) or meant to be semi-regularly revised and switched out. I'm not in the Earthstar community so I don't know what's the norm here.

For the avatar image file for instance, while I think it does make sense to limit it to 1:1 images (for the GUI designer's benefit), the file size and dimensions limits might be out of touch in the future, say with really HiDPI displays or really good compression. However, they seem OK for something that's only meant to be active for 2-4 years.

sgwilym committed:
Soften restrictions regarding image format and contents.

sgwilym commented:
Thanks @parnikkapore, I've modified the language to be less restrictive but still gesture towards some of the concerns interface builders and users may face. 

> To elaborate: I think that this depends on whether the spec is meant to be there unmodified for a long time (like an RFC) or meant to be semi-regularly revised and switched out. I'm not in the Earthstar community so I don't know what's the norm here.

I _think_ we will end up with specs which are changed fairly often and which gracefully handle deviations or older versions. Earthstar as a project does not have much experience with this yet.

sgwilym committed:
Update About spec with status

sgwilym committed:
Add biography section to about spec

sgwilym commented:
I've added new Status and Biography sections to this spec. I think that is more or less everything I'd like in this first version.

achou11 reviewed (approved):
Looks good! Only thing that made me think a little harder is the inclusion of the `status` spec. I understand the use case but wondered if it's necessary to include in this spec as opposed to letting it be something that's defined and implemented in client apps.
